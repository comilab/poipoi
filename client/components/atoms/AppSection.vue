<template lang="pug">
component.rounded.bg-white.my-8.p-4(
  :is="tag",
  data-testid="section"
)
  h3.text-lg(
    v-if="hasTitle",
    data-testid="title"
  )
    slot(name="title") {{ title }}
  slot
</template>

<script lang="ts">
import { defineComponent, computed } from 'nuxt-composition-api'

export default defineComponent({
  props: {
    title: {
      type: String,
      required: false,
      default: null
    },
    tag: {
      type: String,
      required: false,
      default: 'section'
    }
  },
  setup (props, context) {
    const hasTitle = computed(() => {
      return props.title || context.slots.title
    })

    return {
      hasTitle
    }
  }
})
</script>
