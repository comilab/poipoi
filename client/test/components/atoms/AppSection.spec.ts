import { render, RenderOptions, within } from '@testing-library/vue'

import AppSection from '~/components/atoms/AppSection.vue'

describe('components/atoms/AppSection', () => {
  let options: RenderOptions<AppSection>
  let props: any

  beforeEach(() => {
    props = {}

    options = {
      props,
      slots: {
        default: 'default-slot',
        title: 'title-slot'
      }
    }
  })

  test('renders correctly', () => {
    const { container, getByText, getByTestId } = render(AppSection, options)

    expect(container).toMatchSnapshot()
    getByText('default-slot')
    within(getByTestId('title')).getByText('title-slot')
  })

  describe('props', () => {
    describe('title', () => {
      beforeEach(() => {
        props.title = 'title'
        delete options.slots?.title
      })

      test('タイトルとして表示', () => {
        const { container, getByTestId } = render(AppSection, options)

        expect(container).toMatchSnapshot()
        within(getByTestId('title')).getByText('title')
      })

      test('titleもslot#titleもなければタイトルを描画しない', () => {
        props.title = null
        const { container, queryByTestId } = render(AppSection, options)

        expect(container).toMatchSnapshot()
        expect(queryByTestId('title')).not.toBeInTheDocument()
      })
    })

    describe('tag', () => {
      test('デフォルトはsection', () => {
        const { container, queryByTestId } = render(AppSection, options)

        expect(container).toMatchSnapshot()
        expect(queryByTestId('section')!.tagName).toBe('SECTION')
      })

      test('指定があればそのタグで描画', () => {
        props.tag = 'div'
        const { container, queryByTestId } = render(AppSection, options)

        expect(container).toMatchSnapshot()
        expect(queryByTestId('section')).toBeInstanceOf(HTMLDivElement)
      })
    })
  })
})
