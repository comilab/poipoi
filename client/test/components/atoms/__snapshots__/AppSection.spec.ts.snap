// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/atoms/AppSection props tag デフォルトはsection 1`] = `
<div>
  <section
    class="rounded bg-white my-8 p-4"
    data-testid="section"
  >
    <h3
      class="text-lg"
      data-testid="title"
    >
      title-slot
    </h3>
    default-slot
  </section>
</div>
`;

exports[`components/atoms/AppSection props tag 指定があればそのタグで描画 1`] = `
<div>
  <div
    class="rounded bg-white my-8 p-4"
    data-testid="section"
  >
    <h3
      class="text-lg"
      data-testid="title"
    >
      title-slot
    </h3>
    default-slot
  </div>
</div>
`;

exports[`components/atoms/AppSection props title titleもslot#titleもなければタイトルを描画しない 1`] = `
<div>
  <section
    class="rounded bg-white my-8 p-4"
    data-testid="section"
  >
    <!---->
    default-slot
  </section>
</div>
`;

exports[`components/atoms/AppSection props title タイトルとして表示 1`] = `
<div>
  <section
    class="rounded bg-white my-8 p-4"
    data-testid="section"
  >
    <h3
      class="text-lg"
      data-testid="title"
    >
      title
    </h3>
    default-slot
  </section>
</div>
`;

exports[`components/atoms/AppSection renders correctly 1`] = `
<div>
  <section
    class="rounded bg-white my-8 p-4"
    data-testid="section"
  >
    <h3
      class="text-lg"
      data-testid="title"
    >
      title-slot
    </h3>
    default-slot
  </section>
</div>
`;
